src/Text/Pandoc/Writers/OpenDocument.hs

```diff
@@ -235,6 +242,15 @@ writeOpenDocument opts (Pandoc meta blocks) = do
                         meta
   ((body, metadata),s) <- flip runStateT
         defaultWriterState $ do
+           let isHeaderIdent (Header _ (ident,_,_) _) = [ident]
+               isHeaderIdent _ = []
+           modify $ \s -> s{ stHeaderRefs = query isHeaderIdent blocks }
```

Comment:
We iterate over all document elements three times. This could be reduced to a single pass by combining the three identifier lists into a single data type as mentioned above.

Reply:
Combined in what I hope is a correct way - I'm slightly unsure of the method of collecting inlines - but it seems to work based on the unit test.